xla/service/gpu/gpu_replica_reduce_splitter.cc

```diff
@@ -0,0 +1,175 @@
+
+namespace m = match;
+
+template <typename Pattern>
+auto OptionalCopy(HloInstruction **optional_copy, Pattern pattern) {
+  return m::AnyOf<HloInstruction>(m::Copy(optional_copy, pattern),
+                                  std::move(pattern));
+}
+
+StatusOr<bool> ReplicaReduceSplitter::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  bool changed = false;
+  const HloModuleConfig &config = module->config();
+  if (!config.use_spmd_partitioning()) {
+    VLOG(2) << "Unsupported module";
+    return false;
+  }
+  HloInstruction *pid = nullptr;
+  for (HloComputation *computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction *instr : computation->MakeInstructionPostOrder()) {
+      if (instr->opcode() == HloOpcode::kPartitionId) {
+        VLOG(2) << "Found partition-id\n";
+        pid = instr;
+        VLOG(2) << instr->GetModule()->config().num_partitions() << "\n";
+        VLOG(2) << instr->GetModule()->config().replica_count() << "\n";
+        break;
+      }
+    }
+  }
+  if (!pid) {
+    VLOG(2) << "No partition-id found!";
+    return false;
+  }
+  int64_t next_channel_id = hlo_query::NextChannelId(*module);
+  for (HloComputation *computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction *instr : computation->MakeInstructionPostOrder()) {
+      if (instr->opcode() == HloOpcode::kReduce) {
+        // VLOG(2) << "instr:" << instr->ToString() << "\n";
+        HloInstruction *absop = nullptr;
+        HloInstruction *shard_operand = nullptr;
+        HloInstruction *optcpy = nullptr;
+        HloInstruction *operand = instr->mutable_operand(0);
+        if (Match(
+                operand,
+                m::Abs(&absop, OptionalCopy(&optcpy, m::Op(&shard_operand))))) {
+          if (optcpy) VLOG(2) << "optcpy " << optcpy->ToString() << "\n";
+          VLOG(2) << "instr:" << instr->ToString() << "\n";
+          VLOG(2) << "shard_operand:" << shard_operand->ToString() << "\n";
+          // return true;
+          if (shard_operand->opcode() != HloOpcode::kDynamicSlice &&
+              shard_operand->sharding().IsReplicated()) {
+            VLOG(2) << "Found replicated operand\n";
+            int num_partitions = instr->GetModule()->config().num_partitions();
+            // HloInstruction *zero = instr->AddInstruction(
+            //     HloInstruction::CreateConstant(LiteralUtil::Zero(S32)));
+            const Shape target = shard_operand->shape();
+            size_t num_col = target.dimensions(1);
+            size_t num_row = target.dimensions(0);
+            VLOG(2) << "tensor shape: " << num_row << " x " << num_col
+                    << std::endl;
+            Shape ds_shape =
+                ShapeUtil::MakeShape(pid->shape().element_type(), {1});
+            std::vector<int> all_dev(num_partitions);
+            std::iota(std::begin(all_dev), std::end(all_dev), 0);
```

Comment:
absl::c_iota(all_dev, 0) could be used here
(header absl/algorithm/container.h)

Reply:
Fixed.